red_blue_world/Quilt.py

```diff
@@ -29,52 +31,60 @@ def __init__(self) -> None:
     def step(self, a: Action) -> Tuple[AgentState, Reward]:
         s, r, d = self._active_patch.step(a)
 
-        # TODO: instead of relying on direction to signal unloading and transitioning
-        # use a new signal coming from Patch.step to signal that it is time to transition
+        # use direction signal coming from Patch.step to signal that it is time to transition
         if d != Direction.none:
             self._maybe_unload(self._active_patch_id)
-            next_id = self._handle_patch_transition(self._active_patch_id, d)
+            next_id = self._handle_patch_transition(self._active_patch_id, d, s)
 
             self._active_patch_id = next_id
             self._active_patch = self._patches[next_id]
             self._active_patch.on_enter(s)
-
+            
         return (s, r)
 
-    def _handle_patch_transition(self, patch_id: PatchID, d: Direction) -> PatchID:
+    def _handle_patch_transition(self, patch_id: PatchID, d: Direction, agent_loc: AgentState) -> PatchID:
         if d == Direction.up: next_id = _up(patch_id)
         elif d == Direction.down: next_id = _down(patch_id)
         elif d == Direction.left: next_id = _left(patch_id)
         else:
             assert d == Direction.right
             next_id = _right(patch_id)
-
+        
+        next_loc = patch_loader.transit_agent(d, agent_loc)
+        
         # *synchronously* ensure the next patch is loaded
         # if this is anything more than a no-op, we screwed up somewhere
-        self._ensure_load(next_id)
+        self._ensure_load(next_id, next_loc)
 
-        # asynchronously load the 9x9 square around the next patch
-        self._back_thread.submit(self._ensure_load9x9, patch_id)
+        # asynchronously load the 3x3 square around the next patch
+        self._back_thread.submit(self._ensure_load3x3, (patch_id, next_loc))
 
         return next_id
 
-    def _ensure_load(self, patch_id: PatchID) -> None:
+    def _ensure_load(self, patch_id: PatchID, agent_loc: AgentState) -> None:
         # shortcut if there is no work to be done
         if patch_id in self._patches: return
 
         if self.patch_exists(patch_id):
-            patch = self.load_patch(patch_id)
+            patch = self.load_patch(patch_id, agent_loc)
         else:
-            patch = self.build_patch()
+            patch = self.build_patch(agent_loc)
 
         self._patches[patch_id] = patch
 
-    def _ensure_load9x9(self, patch_id: PatchID) -> None:
+    def _ensure_load3x3(self, patch_id: PatchID, agent_loc: AgentState) -> None:
+        x, y = patch_id
+
+        for dx, dy in product(range(-1, 2), range(-1, 2)):
+            coord = (x + dx, y + dy)
+            self._ensure_load(coord, agent_loc)
+
+    def _ensure_load9x9(self, patch_id: PatchID, agent_loc: AgentState) -> None:
```

Comment:
We might want to change this to a `maybe_load9x9` later, since this doesn't need to be a synchronization point with the loader thread.

But that can be a later problem.